### PR TITLE
FakeTensor._normalize_fake_device skips GPU context init for PrivateUse1 backends (#191293)

### PR DESCRIPTION
**Pull request by TangSiLiYang**
## Bug description

`FakeTensor._normalize_fake_device` (torch/_subclasses/fake_tensor.py, L935)
hardcodes a `("cuda", "xpu")` whitelist to decide whether to call
`init_gpu_context(device)`. PrivateUse1 backends (e.g. MLU, NPU) that register
via `torch.utils.rename_privateuse1_backend()` are silently skipped, causing
`FakeTensor` creation to fail or produce incorrect results when the backend
requires GPU context initialization before fake tensor operations.

## Reproduction

```python
import torch

# Register a PrivateUse1 backend (e.g. MLU)
torch.utils.rename_privateuse1_backend("mlu")

# Any path that creates FakeTensors on the MLU device will skip
# init_gpu_context, because "mlu" is not in ("cuda", "xpu")
from torch._subclasses.fake_tensor import FakeTensorMode
mode = FakeTensorMode()
with mode:
    t = torch.empty(4, device="mlu")  # no init_gpu_context called
```

## Root cause

```python
# torch/_subclasses/fake_tensor.py L935
if device.type in ("cuda", "xpu"):
    init_gpu_context(device)
```

The check should also cover the renamed PrivateUse1 backend.

## Proposed fix

```python
if device.type in ("cuda", "xpu") or device.type == torch._C._get_privateuse1_backend_name():
    init_gpu_context(device)
```

`torch._C._get_privateuse1_backend_name()` returns the current PrivateUse1
backend name (default `"privateuseone"`, or the renamed value after
`rename_privateuse1_backend()`). When no custom backend is registered it
returns `"privateuseone"`, which will never match a real device type, so
the default behavior is unchanged.

## Impact

All PrivateUse1 backends that need GPU context initialization (MLU, NPU, etc.)
hit this when using FakeTensor / torch.compile / Dynamo tracing. Downstream
vendors currently work around this with monkey-patches in their own repos.

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -932,7 +932,10 @@ def fake_device(self, device: torch.device) -> None:
     @staticmethod
     def _normalize_fake_device(device: torch.device) -> torch.device:
         """Normalize device by initializing GPU context and setting device index."""
-        if device.type in ("cuda", "xpu"):
+        if (
+            device.type in ("cuda", "xpu")
+            or device.type == torch._C._get_privateuse1_backend_name()
+        ):
             init_gpu_context(device)
```
Comment (aorenste) [inline torch/_subclasses/fake_tensor.py:939]: [blocker] My agent says: `init_gpu_context` is a no-op today, so this call -- and the new PrivateUse1 branch above -- has no observable effect, and the bug this PR targets is not actually fixed. Its guard (line 708) is

    if torch.accelerator.current_accelerator(True) == device.type:

which compares a `torch.device` (what `current_accelerator` returns) against a `str` (`device.type`), so it is always False. Verified on a CUDA box by instrumenting `torch.empty`/`torch.zeros`: the warmup allocation never fires -- for cuda/xpu or PrivateUse1. The guard itself needs fixing:

    if (acc := torch.accelerator.current_accelerator(True)) is not None and acc.type == device.type:

This is pre-existing (introduced by #176409, whose stated intent was "the current accelerator matches device.type" but which compared a device to a string); the `.type == .type` form implements that intent and does not re-break the XPU case #176409 fixed -- on an XPU box, normalizing a `cuda` device stays a no-op.

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -932,7 +932,10 @@ def fake_device(self, device: torch.device) -> None:
     @staticmethod
     def _normalize_fake_device(device: torch.device) -> torch.device:
         """Normalize device by initializing GPU context and setting device index."""
-        if device.type in ("cuda", "xpu"):
+        if (
```
Comment (aorenste) [inline torch/_subclasses/fake_tensor.py:935]: [cleanup] My agent says: this adds a third device-type list that is narrower than and inconsistent with `_is_indexed_device_type` (used just below at line 941), which already lists cuda/hpu/xpu/mps/mtia plus the PrivateUse1 name. Once `init_gpu_context` self-guards on the current accelerator (see the other comment), this whitelist is redundant -- drop it and call `init_gpu_context(device)` unconditionally, which also covers hpu/mps/mtia for free:

    init_gpu_context(device)

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] This bug fix needs a regression test, and it should assert the *effect* -- that the accelerator context actually gets initialized -- not merely that `init_gpu_context` was called (a call-spy passes even while the guard is broken). A registered PrivateUse1 backend becomes the current accelerator (aten/src/ATen/DeviceAccelerator.cpp), so the OpenReg CI backend can drive the exact PrivateUse1 path end-to-end:

    self.assertFalse(torch_openreg.openreg.is_initialized())
    with FakeTensorMode():
        torch.empty(4, device="openreg")
    self.assertTrue(torch_openreg.openreg.is_initialized())

This fails unless both the gate routes PrivateUse1 into `init_gpu_context` and the guard actually fires. Verified on CUDA that the analogous `torch.cuda.is_initialized()` flips False->True once the guard is fixed and stays False with the current guard, so a device-generic variant (`torch.<accel>.is_initialized()`) would catch the same regression on cuda/xpu.
